File: l016/lecture016.cpp
#include <iostream>
#include <string>
#include <vector>
#include <climits>
#include <queue>
#include <unordered_map>
using namespace std;

//PROGRAM 1
//TO FIND Kth LARGEST ELEMENT IN AN ARRAY
void downheapify(vector<int> &arr, int idx, int lidx)
{
	int max_idx = idx;
	int lci = 2 * idx + 1;
	int rci = 2 * idx + 2;

	if (lci <= lidx && arr[lci] > arr[max_idx])
	{
		max_idx = lci;
	}

	if (rci <= lidx && arr[rci] > arr[max_idx])
	{
		max_idx = rci;
	}

	if (max_idx != idx)
	{
		swap(arr[idx], arr[max_idx]);
		downheapify(arr, max_idx, lidx);
	}
}

int findKthLargest(vector<int> &arr, int k)
{
	int n = arr.size() - 1;

	for (int i = n; i >= 0; i--)
	{
		downheapify(arr, i, n);
	}

	for (int i = 0; i < k; i++)
	{
		swap(arr[0], arr[n - i]);
		downheapify(arr, 0, n - i - 1);
	}

	return arr[n - k + 1];
}

//PROGRAM 2
//TO FIND Kth LARGEST ELEMENT IN A STREAM
class KthLargest
{
public:
	priority_queue<int, vector<int>, greater<int>> pq;
	int size;
	KthLargest(int k, vector<int> &nums)
	{
		size = k;

		for (int ele : nums)
		{
			pq.push(ele);

			if (pq.size() > k)
				pq.pop();
		}
	}

	int add(int val)
	{
		pq.push(val);
		if (pq.size() > size)
			pq.pop();

		return pq.top();
	}
};

//PROGRAM 3
//K MOST FREQUENT ELEMENTS IN AN ARRAY REFER LEETOCDE 347 
vector<int> topKFrequent(vector<int> &nums, int k)
{
	unordered_map<int, int> map;
	for (int ele : nums)
		map[ele]++;

	// first -> freq, second -> number and by default it is min PQ.
	priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;

	for (pair<int, int> key : map)
	{
		pq.push({key.second, key.first});
		if (pq.size() > k)
			pq.pop();
	}

	vector<int> ans;
	while (pq.size() != 0)
	{
		ans.push_back(pq.top().second);
		pq.pop();
	}
	return ans;
}

//PROGRAM 4
//FIND MEDIAN IN A DATA STREAM REFER LEETCODE
class MedianFinder {
public:
    priority_queue<int> smallerPQ;                            // max PQ, isme hme chote number daalne hai
															  // jiska top in sabka max ho
															  // iska size n/2 se zyada nhi hona chahiye
															   
    priority_queue<int, vector<int>, greater<int>> greaterPQ; // min PQ, isme hme bade number daalne hai
															  // jiska top in sabka min ho
															  // iska size n/2 ya n/2 + 1 se zyada nhi hona chahiye

public:
    MedianFinder()
    {
    }

    void addNum(int num)
    {
        if (smallerPQ.size() == 0 || smallerPQ.top() > num) // chote number daalna hai
            smallerPQ.push(num);
        else												// bada number daalna hai
            greaterPQ.push(num);

        if (smallerPQ.size() > (greaterPQ.size() + 1))      // smallerPQ ka size maintain karna hai n/2
															// isliye smallerPQ se pop krwa ke greaterPQ mein dala
        {
            greaterPQ.push(smallerPQ.top());
            smallerPQ.pop();
        }
        else if (smallerPQ.size() < greaterPQ.size())       // smallerPQ ka size maintain karna hai n/2
															// isliye greaterPQ se pop krwa ke smallerPQ mein dala
        {
            smallerPQ.push(greaterPQ.top());
            greaterPQ.pop();
        }
    }

    double findMedian()
    {
        if (smallerPQ.size() == greaterPQ.size())
            return smallerPQ.size() == 0 ? 0 : (smallerPQ.top() + greaterPQ.top()) / 2.0;
        else
            return smallerPQ.top();
    }
};

int main(int argc, char **argv)
{
	vector<int> arr = {2, 5, 2, 6, 7, 7, 2, 6, 6, 9, 1, 0, -1, 8};
	cout << findKthLargest(arr, 4);
	return 0;
}

File: l016/lecture016.java
import java.util.PriorityQueue;

public class lecture016 {

    // PROGRAM 1
    // TO FIND Kth LARGEST ELEMENT IN AN ARRAY
    int kthLargest(int[] arr) {
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        int size = 0;

        for (int i = 0; i < n; i++) {
            pq.push(arr[i]);

            if (pq.size() > k) {
                pq.poll();
            }
        }

        return pq.peek();
    }

    // PROGRAM 2
    // TO FIND Kth LARGEST ELEMENT IN A SORTED MATRIX
    public static class Pair {
        int ele, i, j;

        Pair(int ele, int i, int j) {
            this.ele = ele;
            this.i = i;
            this.j = j;
        }
    }

    public static int kth_smallest(int[][] arr, int k) {
        PriorityQueue<Pair> pq = new PriorityQueue<>((Pair a, Pair b) -> {
            return a.ele - b.ele;
        }); // makes min heap

        int n = arr.length;
        int m = arr[0].length;

        for (int i = 0; i < n; i++) {
            pq.add(new Pair(arr[i][0], i, 0));
        }

        while (--k > 0) {
            Pair p = pq.poll();
            int x = p.i;
            int y = p.j;

            if (++y < m)
                pq.add(new Pair(arr[x][y], x, y));
        }

        return pq.poll().ele;
    }

    // PROGRAM 3
    // K MOST FREQUENT ELEMENTS IN AN ARRAY REFER LEETOCDE 347
    public int[] topKFrequent(int[] nums, int k) {
        HashMap<Integer, Integer> map = new HashMap<>();
        for (int ele : nums)
            map.put(ele, map.getOrDefault(ele, 0) + 1);

        // first -> freq, second -> number and by default it is min PQ.
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> {
            return a[0] - b[0];
        });

        for (Integer key : map.keySet()) {
            pq.add(new int[2] { map.get(key), key });
            if (pq.size() > k)
                pq.remove();
        }

        int[] ans = new int[pq.size()];
        int i = 0;
        while (pq.size() != 0) {
            int[] ar = pq.remove();
            ans[i++] = ar[1];
        }

        return ans;
    }

    // PROGRAM 4
    // SWIM IN RISING WATER LEETCODE BFS DIJKSTRA
    public int swimInWater(int[][] grid) {
        int n = grid.length;
        int ans = Math.max(grid[0][0], grid[n - 1][n - 1]);

        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> { // r,c,h
            return a[2] - b[2];
        });

        boolean[][] vis = new boolean[n][n];
        vis[0][0] = true;

        int[][] dir = { { -1, 0 }, { 0, -1 }, { 1, 0 }, { 0, 1 } };
        pq.add(new int[] { 0, 0, grid[0][0] });

        while (pq.size() != 0) {
            int[] vtx = pq.poll();
            ans = Math.max(ans, vtx[2]);

            for (int d = 0; d < 4; d++) {
                int x = vtx[0] + dir[d][0];
                int y = vtx[1] + dir[d][1];

                if (x >= 0 && y >= 0 && x < n && y < n && !vis[x][y]) {
                    if (x == n - 1 && y == n - 1)
                        return ans;
                    pq.add(new int[] { x, y, grid[x][y] });
                    vis[x][y] = true;
                }
            }
        }

        return -1;
    }

    // MAIN
    public static void main(String[] args) {
        int[][] arr = { { 10, 30, 70, 100 }, { 11, 105, 130, 170 }, { 123, 136, 153, 200 } };

        System.out.println(kth_smallest(arr, 5));
    }
}